core/jvm/src/main/scala/scalaz/zio/RTS.scala

```diff
@@ -760,181 +673,130 @@ private object RTS {
+      oldState match {
+        case Executing(interrupted, status, observers0) =>
+          val observers = k :: observers0
 
-        case s @ AsyncRegion(_, _, _, _, _) =>
-          val newStatus = s.copy(observers = cb :: s.observers)
-
-          if (!status.compareAndSet(oldStatus, newStatus)) observe0(cb)
-          else Async.later[Nothing, ExitResult[E, A]]
+          if (!state.compareAndSet(oldState, Executing(interrupted, status, observers))) observe0(k)
+          else Async.later
 
         case Done(v) => Async.now(ExitResult.succeeded(v))
       }
     }
 
-    private final def tryObserve0: Option[ExitResult[E, A]] =
-      status.get match {
+    private final def poll0: Option[ExitResult[E, A]] =
+      state.get match {
         case Done(r) => Some(r)
         case _       => None
       }
 
-    private final def purgeObservers(v: ExitResult[E, A], observers: List[Callback[Nothing, ExitResult[E, A]]]): Unit =
+    private final def notifyObservers(
+      v: ExitResult[E, A],
+      observers: List[Callback[Nothing, ExitResult[E, A]]]
+    ): Unit = {
+      val result = ExitResult.succeeded(v)
+
       // To preserve fair scheduling, we submit all resumptions on the thread
       // pool in order of their submission.
-      observers.reverse.foreach(k => rts.submit(k(ExitResult.succeeded(v))))
+      observers.reverse.foreach(k => rts.submit(k(result)))
+    }
+  }
+  sealed abstract class FiberStatus extends Serializable with Product {
+    final def suspended: Boolean = this match {
+      case FiberStatus.Running      => false
+      case FiberStatus.Suspended(_) => true
+    }
+    def cancel: Option[Canceler]
+  }
+  object FiberStatus {
+    final case object Running extends FiberStatus {
+      final def cancel: Option[Canceler] = None
+    }
+    final case class Suspended(cancel: Option[IO[Nothing, Unit]]) extends FiberStatus
   }
 
-  sealed abstract class FiberStatus[E, A] extends Serializable with Product {
+  sealed abstract class FiberState[E, A] extends Serializable with Product {
 
     /** indicates if the fiber was interrupted */
     def interrupted: Boolean
 
   }
-  object FiberStatus extends Serializable {
+  object FiberState extends Serializable {
     final case class Executing[E, A](
       interrupted: Boolean,
+      status: FiberStatus,
       observers: List[Callback[Nothing, ExitResult[E, A]]]
-    ) extends FiberStatus[E, A]
-    final case class AsyncRegion[E, A](
-      interrupted: Boolean,
-      reentrancy: Int,
-      resume: Int,
-      cancel: Option[Canceler],
-      observers: List[Callback[Nothing, ExitResult[E, A]]]
-    ) extends FiberStatus[E, A]
-    final case class Done[E, A](value: ExitResult[E, A]) extends FiberStatus[E, A] {
-      override def interrupted: Boolean = false
+    ) extends FiberState[E, A]
+    final case class Done[E, A](value: ExitResult[E, A]) extends FiberState[E, A] {
+      def interrupted: Boolean = value.interrupted
```

Comment:
If we use `val`, it will store memory for `interrupted` even if it's not used. But we can recompute it on the fly if it's called, so I think it's better to use `def` to save memory (calling it will be uncommon).